# BioLink API
#
# API integration layer for linked biological objects.   __Source:__ https://github.com/biolink/biolink-api/
#
# OpenAPI spec version: 0.1.1
# 
# Generated by: https://github.com/swagger-api/swagger-codegen.git

#' @title Bioentity operations
#' @description swagger.Bioentity
#'
#' @field path Stores url path of the request.
#' @field apiClient Handles the client-server communication.
#' @field userAgent Set the user agent of the request.
#'
#' @importFrom R6 R6Class
#'
#' @section Methods:
#' \describe{
#'
#' get_allele_object Returns genotype object
#'
#'
#' get_anatomy_gene_associations Returns expression events for a gene
#'
#'
#' get_anatomy_gene_by_taxon_associations Returns gene ids for all genes for a particular anatomy in a taxon
#'
#'
#' get_anatomy_object Returns basic info on anatomical entity
#'
#'
#' get_disease_gene_associations Returns genes associated with a disease
#'
#'
#' get_disease_model_associations Returns associations to models of the disease
#'
#'
#' get_disease_model_taxon_associations Returns associations to models of the disease constrained by taxon
#'
#'
#' get_disease_object Returns basic info on a disease
#'
#'
#' get_disease_phenotype_associations Returns phenotypes associated with disease
#'
#'
#' get_disease_substance_associations Returns substances associated with a disease
#'
#'
#' get_gene_disease_associations Returns diseases associated with gene
#'
#'
#' get_gene_expression_associations Returns expression events for a gene
#'
#'
#' get_gene_function_associations Returns function associations for a gene
#'
#'
#' get_gene_homolog_associations Returns homologs for a gene
#'
#'
#' get_gene_interactions Returns interactions for a gene
#'
#'
#' get_gene_object Returns basic info about a gene
#'
#'
#' get_gene_pathway_associations Returns pathways associated with gene
#'
#'
#' get_gene_phenotype_associations Returns phenotypes associated with gene
#'
#'
#' get_generic_associations Returns associations for an entity regardless of the type
#'
#'
#' get_generic_object Returns basic info on object of any type
#'
#'
#' get_genotype_disease_associations Returns diseases associated with a genotype
#'
#'
#' get_genotype_gene_associations Returns genes associated with a genotype
#'
#'
#' get_genotype_genotype_associations Returns genotypes-genotype associations
#'
#'
#' get_genotype_object Returns genotype object
#'
#'
#' get_genotype_phenotype_associations Returns phenotypes associated with a genotype
#'
#'
#' get_goterm_gene_associations Returns associations to GO terms for a gene
#'
#'
#' get_goterm_object Returns basic info on a gene
#'
#'
#' get_parent_object Returns basic info on an individual/case
#'
#'
#' get_pathway_gene_associations Returns genes associated with a pathway
#'
#'
#' get_pathway_object Returns basic info on a pathway
#'
#'
#' get_phenotype_anatomy_associations Returns anatomical entities associated with a phenotype
#'
#'
#' get_phenotype_function_associations Returns biological functions associated with a Phenotype
#'
#'
#' get_phenotype_gene_associations Returns associated phenotypes
#'
#'
#' get_phenotype_gene_by_taxon_associations Returns gene ids for all genes for a particular phenotype in a taxon
#'
#'
#' get_phenotype_object Returns basic info on a phenotype
#'
#'
#' get_substance_object Returns basic info on a substance
#'
#'
#' get_substance_participant_in_associations Returns associations between an activity and process and the specified substance
#'
#'
#' get_substance_role_associations Returns associations between given drug and roles
#'
#'
#' get_substance_treats_associations Returns substances associated with a disease
#'
#'
#' get_variant_gene_associations Returns genes associated with a variant
#'
#'
#' get_variant_genotype_associations Returns genotypes associated with a variant
#'
#'
#' get_variant_object Returns basic info on a variant
#'
#'
#' get_variant_phenotype_associations Returns phenotypes associated with a variant
#'
#' }
#'
#' @export
BioentityApi <- R6::R6Class(
  'BioentityApi',
  public = list(
    userAgent = "Swagger-Codegen/1.0.0/r",
    apiClient = NULL,
    initialize = function(apiClient){
      if (!missing(apiClient)) {
        self$apiClient <- apiClient
      }
      else {
        self$apiClient <- ApiClient$new()
      }
    },
    get_allele_object = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/allele/{id}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- Genotype$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_anatomy_gene_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/anatomy/{id}/genes/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_anatomy_gene_by_taxon_associations = function(id, taxid, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/anatomy/{id}/genes/{taxid}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      if (!missing(`taxid`)) {
        urlPath <- gsub(paste0("\\{", "taxid", "\\}"), `taxid`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        # void response, no need to return anything
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_anatomy_object = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/anatomy/{id}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- BioObject$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_disease_gene_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/disease/{id}/genes/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_disease_model_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/disease/{id}/models/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_disease_model_taxon_associations = function(id, taxon, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/disease/{id}/models/{taxon}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      if (!missing(`taxon`)) {
        urlPath <- gsub(paste0("\\{", "taxon", "\\}"), `taxon`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_disease_object = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/disease/{id}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- BioObject$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_disease_phenotype_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/disease/{id}/phenotypes/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_disease_substance_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/disease/{id}/treatment/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        # void response, no need to return anything
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_gene_disease_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/gene/{id}/diseases/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_gene_expression_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/gene/{id}/expression/anatomy"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_gene_function_associations = function(id, start, exclude_automatic_assertions, rows, homolog_taxon, homology_type, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`homolog_taxon`)) {
        queryParams['homolog_taxon'] <- homolog_taxon
      }

      if (!missing(`homology_type`)) {
        queryParams['homology_type'] <- homology_type
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/gene/{id}/function/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_gene_homolog_associations = function(id, homology_type, homolog_taxon, rows, start, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, exclude_automatic_assertions, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`homology_type`)) {
        queryParams['homology_type'] <- homology_type
      }

      if (!missing(`homolog_taxon`)) {
        queryParams['homolog_taxon'] <- homolog_taxon
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      urlPath <- "/bioentity/gene/{id}/homologs/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_gene_interactions = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/gene/{id}/interactions/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_gene_object = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/gene/{id}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- Gene$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_gene_pathway_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/gene/{id}/pathways/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_gene_phenotype_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/gene/{id}/phenotypes/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_generic_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/{id}/associations/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_generic_object = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/{id}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- BioObject$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_genotype_disease_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/genotype/{id}/diseases/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_genotype_gene_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/genotype/{id}/genes/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_genotype_genotype_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/genotype/{id}/genotypes/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_genotype_object = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/genotype/{id}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- Genotype$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_genotype_phenotype_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/genotype/{id}/phenotypes/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_goterm_gene_associations = function(id, relationship_type, exclude_automatic_assertions, rows, start, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`relationship_type`)) {
        queryParams['relationship_type'] <- relationship_type
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/goterm/{id}/genes/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_goterm_object = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/goterm/{id}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- BioObject$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_parent_object = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/individual/{id}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- BioObject$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_pathway_gene_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/pathway/{id}/genes/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_pathway_object = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/pathway/{id}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- BioObject$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_phenotype_anatomy_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/phenotype/{id}/anatomy/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- NamedObject$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_phenotype_function_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/phenotype/{id}/function/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- Association$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_phenotype_gene_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/phenotype/{id}/genes/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_phenotype_gene_by_taxon_associations = function(id, taxid, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/phenotype/{id}/gene/{taxid}/ids"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      if (!missing(`taxid`)) {
        urlPath <- gsub(paste0("\\{", "taxid", "\\}"), `taxid`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        # void response, no need to return anything
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_phenotype_object = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/phenotype/{id}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- BioObject$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_substance_object = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/substance/{id}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- BioObject$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_substance_participant_in_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/substance/{id}/participant_in/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- Association$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_substance_role_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/substance/{id}/roles/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- Association$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_substance_treats_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/substance/{id}/treats/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        # void response, no need to return anything
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_variant_gene_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/variant/{id}/genes/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_variant_genotype_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/variant/{id}/genotypes/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_variant_object = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/variant/{id}"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- BioObject$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    },
    get_variant_phenotype_associations = function(id, start, exclude_automatic_assertions, rows, evidence, slim, use_compact_associations, fetch_objects, unselect_evidence, ...){
      args <- list(...)
      queryParams <- list()
      headerParams <- character()

      if (!missing(`start`)) {
        queryParams['start'] <- start
      }

      if (!missing(`exclude_automatic_assertions`)) {
        queryParams['exclude_automatic_assertions'] <- exclude_automatic_assertions
      }

      if (!missing(`rows`)) {
        queryParams['rows'] <- rows
      }

      if (!missing(`evidence`)) {
        queryParams['evidence'] <- evidence
      }

      if (!missing(`slim`)) {
        queryParams['slim'] <- slim
      }

      if (!missing(`use_compact_associations`)) {
        queryParams['use_compact_associations'] <- use_compact_associations
      }

      if (!missing(`fetch_objects`)) {
        queryParams['fetch_objects'] <- fetch_objects
      }

      if (!missing(`unselect_evidence`)) {
        queryParams['unselect_evidence'] <- unselect_evidence
      }

      urlPath <- "/bioentity/variant/{id}/phenotypes/"
      if (!missing(`id`)) {
        urlPath <- gsub(paste0("\\{", "id", "\\}"), `id`, urlPath)
      }

      resp <- self$apiClient$callApi(url = paste0(self$apiClient$basePath, urlPath),
                                 method = "GET",
                                 queryParams = queryParams,
                                 headerParams = headerParams,
                                 body = body,
                                 ...)
      
      if (httr::status_code(resp) >= 200 && httr::status_code(resp) <= 299) {
        returnObject <- AssociationResults$new()
        result <- returnObject$fromJSON(httr::content(resp, "text", encoding = "UTF-8"))
        Response$new(returnObject, resp)
      } else if (httr::status_code(resp) >= 400 && httr::status_code(resp) <= 499) {
        Response$new("API client error", resp)
      } else if (httr::status_code(resp) >= 500 && httr::status_code(resp) <= 599) {
        Response$new("API server error", resp)
      }

    }
  )
)
