# BioLink API
#
# API integration layer for linked biological objects.   __Source:__ https://github.com/biolink/biolink-api/
#
# OpenAPI spec version: 0.1.1
# 
# Generated by: https://github.com/swagger-api/swagger-codegen.git


#' Gene Class
#'
#' @field id 
#' @field label 
#' @field categories 
#' @field types 
#' @field description 
#' @field consider 
#' @field deprecated 
#' @field replaced_by 
#' @field synonyms 
#' @field xrefs 
#' @field taxon 
#' @field literature_associations 
#' @field summaries 
#' @field disease_associations 
#' @field full_name 
#' @field phenotype_associations 
#' @field families 
#' @field function_associations 
#' @field chromosome 
#' @field genotype_associations 
#' @field systematic_name 
#' @field sequence_features 
#' @field homology_associations 
#' @field interaction_associations 
#' @field transcripts 
#' @field pathway_associations 
#'
#' @importFrom R6 R6Class
#' @importFrom jsonlite fromJSON toJSON
#' @export
Gene <- R6::R6Class(
  'Gene',
  public = list(
    `id` = NULL,
    `label` = NULL,
    `categories` = NULL,
    `types` = NULL,
    `description` = NULL,
    `consider` = NULL,
    `deprecated` = NULL,
    `replaced_by` = NULL,
    `synonyms` = NULL,
    `xrefs` = NULL,
    `taxon` = NULL,
    `literature_associations` = NULL,
    `summaries` = NULL,
    `disease_associations` = NULL,
    `full_name` = NULL,
    `phenotype_associations` = NULL,
    `families` = NULL,
    `function_associations` = NULL,
    `chromosome` = NULL,
    `genotype_associations` = NULL,
    `systematic_name` = NULL,
    `sequence_features` = NULL,
    `homology_associations` = NULL,
    `interaction_associations` = NULL,
    `transcripts` = NULL,
    `pathway_associations` = NULL,
    initialize = function(`id`, `label`, `categories`, `types`, `description`, `consider`, `deprecated`, `replaced_by`, `synonyms`, `xrefs`, `taxon`, `literature_associations`, `summaries`, `disease_associations`, `full_name`, `phenotype_associations`, `families`, `function_associations`, `chromosome`, `genotype_associations`, `systematic_name`, `sequence_features`, `homology_associations`, `interaction_associations`, `transcripts`, `pathway_associations`){
      if (!missing(`id`)) {
        stopifnot(is.character(`id`), length(`id`) == 1)
        self$`id` <- `id`
      }
      if (!missing(`label`)) {
        stopifnot(is.character(`label`), length(`label`) == 1)
        self$`label` <- `label`
      }
      if (!missing(`categories`)) {
        stopifnot(is.list(`categories`), length(`categories`) != 0)
        lapply(`categories`, function(x) stopifnot(is.character(x)))
        self$`categories` <- `categories`
      }
      if (!missing(`types`)) {
        stopifnot(is.list(`types`), length(`types`) != 0)
        lapply(`types`, function(x) stopifnot(is.character(x)))
        self$`types` <- `types`
      }
      if (!missing(`description`)) {
        stopifnot(is.character(`description`), length(`description`) == 1)
        self$`description` <- `description`
      }
      if (!missing(`consider`)) {
        stopifnot(is.list(`consider`), length(`consider`) != 0)
        lapply(`consider`, function(x) stopifnot(is.character(x)))
        self$`consider` <- `consider`
      }
      if (!missing(`deprecated`)) {
        self$`deprecated` <- `deprecated`
      }
      if (!missing(`replaced_by`)) {
        stopifnot(is.list(`replaced_by`), length(`replaced_by`) != 0)
        lapply(`replaced_by`, function(x) stopifnot(is.character(x)))
        self$`replaced_by` <- `replaced_by`
      }
      if (!missing(`synonyms`)) {
        stopifnot(is.list(`synonyms`), length(`synonyms`) != 0)
        lapply(`synonyms`, function(x) stopifnot(R6::is.R6(x)))
        self$`synonyms` <- `synonyms`
      }
      if (!missing(`xrefs`)) {
        stopifnot(is.list(`xrefs`), length(`xrefs`) != 0)
        lapply(`xrefs`, function(x) stopifnot(is.character(x)))
        self$`xrefs` <- `xrefs`
      }
      if (!missing(`taxon`)) {
        stopifnot(R6::is.R6(`taxon`))
        self$`taxon` <- `taxon`
      }
      if (!missing(`literature_associations`)) {
        stopifnot(is.list(`literature_associations`), length(`literature_associations`) != 0)
        lapply(`literature_associations`, function(x) stopifnot(R6::is.R6(x)))
        self$`literature_associations` <- `literature_associations`
      }
      if (!missing(`summaries`)) {
        stopifnot(is.list(`summaries`), length(`summaries`) != 0)
        lapply(`summaries`, function(x) stopifnot(R6::is.R6(x)))
        self$`summaries` <- `summaries`
      }
      if (!missing(`disease_associations`)) {
        stopifnot(is.list(`disease_associations`), length(`disease_associations`) != 0)
        lapply(`disease_associations`, function(x) stopifnot(R6::is.R6(x)))
        self$`disease_associations` <- `disease_associations`
      }
      if (!missing(`full_name`)) {
        stopifnot(is.character(`full_name`), length(`full_name`) == 1)
        self$`full_name` <- `full_name`
      }
      if (!missing(`phenotype_associations`)) {
        stopifnot(is.list(`phenotype_associations`), length(`phenotype_associations`) != 0)
        lapply(`phenotype_associations`, function(x) stopifnot(R6::is.R6(x)))
        self$`phenotype_associations` <- `phenotype_associations`
      }
      if (!missing(`families`)) {
        stopifnot(is.list(`families`), length(`families`) != 0)
        lapply(`families`, function(x) stopifnot(R6::is.R6(x)))
        self$`families` <- `families`
      }
      if (!missing(`function_associations`)) {
        stopifnot(is.list(`function_associations`), length(`function_associations`) != 0)
        lapply(`function_associations`, function(x) stopifnot(R6::is.R6(x)))
        self$`function_associations` <- `function_associations`
      }
      if (!missing(`chromosome`)) {
        stopifnot(R6::is.R6(`chromosome`))
        self$`chromosome` <- `chromosome`
      }
      if (!missing(`genotype_associations`)) {
        stopifnot(is.list(`genotype_associations`), length(`genotype_associations`) != 0)
        lapply(`genotype_associations`, function(x) stopifnot(R6::is.R6(x)))
        self$`genotype_associations` <- `genotype_associations`
      }
      if (!missing(`systematic_name`)) {
        stopifnot(is.character(`systematic_name`), length(`systematic_name`) == 1)
        self$`systematic_name` <- `systematic_name`
      }
      if (!missing(`sequence_features`)) {
        stopifnot(is.list(`sequence_features`), length(`sequence_features`) != 0)
        lapply(`sequence_features`, function(x) stopifnot(R6::is.R6(x)))
        self$`sequence_features` <- `sequence_features`
      }
      if (!missing(`homology_associations`)) {
        stopifnot(is.list(`homology_associations`), length(`homology_associations`) != 0)
        lapply(`homology_associations`, function(x) stopifnot(R6::is.R6(x)))
        self$`homology_associations` <- `homology_associations`
      }
      if (!missing(`interaction_associations`)) {
        stopifnot(is.list(`interaction_associations`), length(`interaction_associations`) != 0)
        lapply(`interaction_associations`, function(x) stopifnot(R6::is.R6(x)))
        self$`interaction_associations` <- `interaction_associations`
      }
      if (!missing(`transcripts`)) {
        stopifnot(is.list(`transcripts`), length(`transcripts`) != 0)
        lapply(`transcripts`, function(x) stopifnot(R6::is.R6(x)))
        self$`transcripts` <- `transcripts`
      }
      if (!missing(`pathway_associations`)) {
        stopifnot(is.list(`pathway_associations`), length(`pathway_associations`) != 0)
        lapply(`pathway_associations`, function(x) stopifnot(R6::is.R6(x)))
        self$`pathway_associations` <- `pathway_associations`
      }
    },
    toJSON = function() {
      GeneObject <- list()
      if (!is.null(self$`id`)) {
        GeneObject[['id']] <- self$`id`
      }
      if (!is.null(self$`label`)) {
        GeneObject[['label']] <- self$`label`
      }
      if (!is.null(self$`categories`)) {
        GeneObject[['categories']] <- self$`categories`
      }
      if (!is.null(self$`types`)) {
        GeneObject[['types']] <- self$`types`
      }
      if (!is.null(self$`description`)) {
        GeneObject[['description']] <- self$`description`
      }
      if (!is.null(self$`consider`)) {
        GeneObject[['consider']] <- self$`consider`
      }
      if (!is.null(self$`deprecated`)) {
        GeneObject[['deprecated']] <- self$`deprecated`
      }
      if (!is.null(self$`replaced_by`)) {
        GeneObject[['replaced_by']] <- self$`replaced_by`
      }
      if (!is.null(self$`synonyms`)) {
        GeneObject[['synonyms']] <- lapply(self$`synonyms`, function(x) x$toJSON())
      }
      if (!is.null(self$`xrefs`)) {
        GeneObject[['xrefs']] <- self$`xrefs`
      }
      if (!is.null(self$`taxon`)) {
        GeneObject[['taxon']] <- self$`taxon`$toJSON()
      }
      if (!is.null(self$`literature_associations`)) {
        GeneObject[['literature_associations']] <- lapply(self$`literature_associations`, function(x) x$toJSON())
      }
      if (!is.null(self$`summaries`)) {
        GeneObject[['summaries']] <- lapply(self$`summaries`, function(x) x$toJSON())
      }
      if (!is.null(self$`disease_associations`)) {
        GeneObject[['disease_associations']] <- lapply(self$`disease_associations`, function(x) x$toJSON())
      }
      if (!is.null(self$`full_name`)) {
        GeneObject[['full_name']] <- self$`full_name`
      }
      if (!is.null(self$`phenotype_associations`)) {
        GeneObject[['phenotype_associations']] <- lapply(self$`phenotype_associations`, function(x) x$toJSON())
      }
      if (!is.null(self$`families`)) {
        GeneObject[['families']] <- lapply(self$`families`, function(x) x$toJSON())
      }
      if (!is.null(self$`function_associations`)) {
        GeneObject[['function_associations']] <- lapply(self$`function_associations`, function(x) x$toJSON())
      }
      if (!is.null(self$`chromosome`)) {
        GeneObject[['chromosome']] <- self$`chromosome`$toJSON()
      }
      if (!is.null(self$`genotype_associations`)) {
        GeneObject[['genotype_associations']] <- lapply(self$`genotype_associations`, function(x) x$toJSON())
      }
      if (!is.null(self$`systematic_name`)) {
        GeneObject[['systematic_name']] <- self$`systematic_name`
      }
      if (!is.null(self$`sequence_features`)) {
        GeneObject[['sequence_features']] <- lapply(self$`sequence_features`, function(x) x$toJSON())
      }
      if (!is.null(self$`homology_associations`)) {
        GeneObject[['homology_associations']] <- lapply(self$`homology_associations`, function(x) x$toJSON())
      }
      if (!is.null(self$`interaction_associations`)) {
        GeneObject[['interaction_associations']] <- lapply(self$`interaction_associations`, function(x) x$toJSON())
      }
      if (!is.null(self$`transcripts`)) {
        GeneObject[['transcripts']] <- lapply(self$`transcripts`, function(x) x$toJSON())
      }
      if (!is.null(self$`pathway_associations`)) {
        GeneObject[['pathway_associations']] <- lapply(self$`pathway_associations`, function(x) x$toJSON())
      }

      GeneObject
    },
    fromJSON = function(GeneJson) {
      GeneObject <- jsonlite::fromJSON(GeneJson)
      if (!is.null(GeneObject$`id`)) {
        self$`id` <- GeneObject$`id`
      }
      if (!is.null(GeneObject$`label`)) {
        self$`label` <- GeneObject$`label`
      }
      if (!is.null(GeneObject$`categories`)) {
        self$`categories` <- GeneObject$`categories`
      }
      if (!is.null(GeneObject$`types`)) {
        self$`types` <- GeneObject$`types`
      }
      if (!is.null(GeneObject$`description`)) {
        self$`description` <- GeneObject$`description`
      }
      if (!is.null(GeneObject$`consider`)) {
        self$`consider` <- GeneObject$`consider`
      }
      if (!is.null(GeneObject$`deprecated`)) {
        self$`deprecated` <- GeneObject$`deprecated`
      }
      if (!is.null(GeneObject$`replaced_by`)) {
        self$`replaced_by` <- GeneObject$`replaced_by`
      }
      if (!is.null(GeneObject$`synonyms`)) {
        self$`synonyms` <- lapply(GeneObject$`synonyms`, function(x) {
          synonymsObject <- SynonymPropertyValue$new()
          synonymsObject$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE))
          synonymsObject
        })
      }
      if (!is.null(GeneObject$`xrefs`)) {
        self$`xrefs` <- GeneObject$`xrefs`
      }
      if (!is.null(GeneObject$`taxon`)) {
        taxonObject <- Taxon$new()
        taxonObject$fromJSON(jsonlite::toJSON(GeneObject$taxon, auto_unbox = TRUE))
        self$`taxon` <- taxonObject
      }
      if (!is.null(GeneObject$`literature_associations`)) {
        self$`literature_associations` <- lapply(GeneObject$`literature_associations`, function(x) {
          literature_associationsObject <- Association$new()
          literature_associationsObject$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE))
          literature_associationsObject
        })
      }
      if (!is.null(GeneObject$`summaries`)) {
        self$`summaries` <- lapply(GeneObject$`summaries`, function(x) {
          summariesObject <- SummaryPropertyValue$new()
          summariesObject$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE))
          summariesObject
        })
      }
      if (!is.null(GeneObject$`disease_associations`)) {
        self$`disease_associations` <- lapply(GeneObject$`disease_associations`, function(x) {
          disease_associationsObject <- Association$new()
          disease_associationsObject$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE))
          disease_associationsObject
        })
      }
      if (!is.null(GeneObject$`full_name`)) {
        self$`full_name` <- GeneObject$`full_name`
      }
      if (!is.null(GeneObject$`phenotype_associations`)) {
        self$`phenotype_associations` <- lapply(GeneObject$`phenotype_associations`, function(x) {
          phenotype_associationsObject <- Association$new()
          phenotype_associationsObject$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE))
          phenotype_associationsObject
        })
      }
      if (!is.null(GeneObject$`families`)) {
        self$`families` <- lapply(GeneObject$`families`, function(x) {
          familiesObject <- NamedObject$new()
          familiesObject$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE))
          familiesObject
        })
      }
      if (!is.null(GeneObject$`function_associations`)) {
        self$`function_associations` <- lapply(GeneObject$`function_associations`, function(x) {
          function_associationsObject <- Association$new()
          function_associationsObject$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE))
          function_associationsObject
        })
      }
      if (!is.null(GeneObject$`chromosome`)) {
        chromosomeObject <- Chromosome$new()
        chromosomeObject$fromJSON(jsonlite::toJSON(GeneObject$chromosome, auto_unbox = TRUE))
        self$`chromosome` <- chromosomeObject
      }
      if (!is.null(GeneObject$`genotype_associations`)) {
        self$`genotype_associations` <- lapply(GeneObject$`genotype_associations`, function(x) {
          genotype_associationsObject <- Association$new()
          genotype_associationsObject$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE))
          genotype_associationsObject
        })
      }
      if (!is.null(GeneObject$`systematic_name`)) {
        self$`systematic_name` <- GeneObject$`systematic_name`
      }
      if (!is.null(GeneObject$`sequence_features`)) {
        self$`sequence_features` <- lapply(GeneObject$`sequence_features`, function(x) {
          sequence_featuresObject <- SequenceFeature$new()
          sequence_featuresObject$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE))
          sequence_featuresObject
        })
      }
      if (!is.null(GeneObject$`homology_associations`)) {
        self$`homology_associations` <- lapply(GeneObject$`homology_associations`, function(x) {
          homology_associationsObject <- Association$new()
          homology_associationsObject$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE))
          homology_associationsObject
        })
      }
      if (!is.null(GeneObject$`interaction_associations`)) {
        self$`interaction_associations` <- lapply(GeneObject$`interaction_associations`, function(x) {
          interaction_associationsObject <- Association$new()
          interaction_associationsObject$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE))
          interaction_associationsObject
        })
      }
      if (!is.null(GeneObject$`transcripts`)) {
        self$`transcripts` <- lapply(GeneObject$`transcripts`, function(x) {
          transcriptsObject <- Transcript$new()
          transcriptsObject$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE))
          transcriptsObject
        })
      }
      if (!is.null(GeneObject$`pathway_associations`)) {
        self$`pathway_associations` <- lapply(GeneObject$`pathway_associations`, function(x) {
          pathway_associationsObject <- Association$new()
          pathway_associationsObject$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE))
          pathway_associationsObject
        })
      }
    },
    toJSONString = function() {
       sprintf(
        '{
           "id": %s,
           "label": %s,
           "categories": [%s],
           "types": [%s],
           "description": %s,
           "consider": [%s],
           "deprecated": %s,
           "replaced_by": [%s],
           "synonyms": [%s],
           "xrefs": [%s],
           "taxon": %s,
           "literature_associations": [%s],
           "summaries": [%s],
           "disease_associations": [%s],
           "full_name": %s,
           "phenotype_associations": [%s],
           "families": [%s],
           "function_associations": [%s],
           "chromosome": %s,
           "genotype_associations": [%s],
           "systematic_name": %s,
           "sequence_features": [%s],
           "homology_associations": [%s],
           "interaction_associations": [%s],
           "transcripts": [%s],
           "pathway_associations": [%s]
        }',
        self$`id`,
        self$`label`,
        lapply(self$`categories`, function(x) paste(paste0('"', x, '"'), sep=",")),
        lapply(self$`types`, function(x) paste(paste0('"', x, '"'), sep=",")),
        self$`description`,
        lapply(self$`consider`, function(x) paste(paste0('"', x, '"'), sep=",")),
        self$`deprecated`,
        lapply(self$`replaced_by`, function(x) paste(paste0('"', x, '"'), sep=",")),
        lapply(self$`synonyms`, function(x) paste(x$toJSON(), sep=",")),
        lapply(self$`xrefs`, function(x) paste(paste0('"', x, '"'), sep=",")),
        self$`taxon`$toJSON(),
        lapply(self$`literature_associations`, function(x) paste(x$toJSON(), sep=",")),
        lapply(self$`summaries`, function(x) paste(x$toJSON(), sep=",")),
        lapply(self$`disease_associations`, function(x) paste(x$toJSON(), sep=",")),
        self$`full_name`,
        lapply(self$`phenotype_associations`, function(x) paste(x$toJSON(), sep=",")),
        lapply(self$`families`, function(x) paste(x$toJSON(), sep=",")),
        lapply(self$`function_associations`, function(x) paste(x$toJSON(), sep=",")),
        self$`chromosome`$toJSON(),
        lapply(self$`genotype_associations`, function(x) paste(x$toJSON(), sep=",")),
        self$`systematic_name`,
        lapply(self$`sequence_features`, function(x) paste(x$toJSON(), sep=",")),
        lapply(self$`homology_associations`, function(x) paste(x$toJSON(), sep=",")),
        lapply(self$`interaction_associations`, function(x) paste(x$toJSON(), sep=",")),
        lapply(self$`transcripts`, function(x) paste(x$toJSON(), sep=",")),
        lapply(self$`pathway_associations`, function(x) paste(x$toJSON(), sep=","))
      )
    },
    fromJSONString = function(GeneJson) {
      GeneObject <- jsonlite::fromJSON(GeneJson)
      self$`id` <- GeneObject$`id`
      self$`label` <- GeneObject$`label`
      self$`categories` <- GeneObject$`categories`
      self$`types` <- GeneObject$`types`
      self$`description` <- GeneObject$`description`
      self$`consider` <- GeneObject$`consider`
      self$`deprecated` <- GeneObject$`deprecated`
      self$`replaced_by` <- GeneObject$`replaced_by`
      self$`synonyms` <- lapply(GeneObject$`synonyms`, function(x) SynonymPropertyValue$new()$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE)))
      self$`xrefs` <- GeneObject$`xrefs`
      TaxonObject <- Taxon$new()
      self$`taxon` <- TaxonObject$fromJSON(jsonlite::toJSON(GeneObject$taxon, auto_unbox = TRUE))
      self$`literature_associations` <- lapply(GeneObject$`literature_associations`, function(x) Association$new()$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE)))
      self$`summaries` <- lapply(GeneObject$`summaries`, function(x) SummaryPropertyValue$new()$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE)))
      self$`disease_associations` <- lapply(GeneObject$`disease_associations`, function(x) Association$new()$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE)))
      self$`full_name` <- GeneObject$`full_name`
      self$`phenotype_associations` <- lapply(GeneObject$`phenotype_associations`, function(x) Association$new()$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE)))
      self$`families` <- lapply(GeneObject$`families`, function(x) NamedObject$new()$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE)))
      self$`function_associations` <- lapply(GeneObject$`function_associations`, function(x) Association$new()$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE)))
      ChromosomeObject <- Chromosome$new()
      self$`chromosome` <- ChromosomeObject$fromJSON(jsonlite::toJSON(GeneObject$chromosome, auto_unbox = TRUE))
      self$`genotype_associations` <- lapply(GeneObject$`genotype_associations`, function(x) Association$new()$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE)))
      self$`systematic_name` <- GeneObject$`systematic_name`
      self$`sequence_features` <- lapply(GeneObject$`sequence_features`, function(x) SequenceFeature$new()$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE)))
      self$`homology_associations` <- lapply(GeneObject$`homology_associations`, function(x) Association$new()$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE)))
      self$`interaction_associations` <- lapply(GeneObject$`interaction_associations`, function(x) Association$new()$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE)))
      self$`transcripts` <- lapply(GeneObject$`transcripts`, function(x) Transcript$new()$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE)))
      self$`pathway_associations` <- lapply(GeneObject$`pathway_associations`, function(x) Association$new()$fromJSON(jsonlite::toJSON(x, auto_unbox = TRUE)))
    }
  )
)
